birdbox/common/auth.py

```diff
@@ -0,0 +1,46 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+from django.conf import settings
+from django.core.exceptions import SuspiciousOperation
+
+from mozilla_django_oidc.auth import OIDCAuthenticationBackend
+
+
+class BirdboxOIDCAuthenticationBackend(OIDCAuthenticationBackend):
+    """Subclass of the main OIDC auth backend that bootstraps any expected
+    admin users from env vars, so that we can set up a Birdbox site
+    without needing to shell into a running container."""
+
```

Comment:
This idea/approach is not perfect by any means. I wanted to solve the problem of 'how do we set up at least one admin user on initial project setup without having to shell in to the running infra', trying to keep to the low-touch/turnkey aim of Birdbox going.

What's below will run every time someone signs in the backend, which will work but isn't great, because it's wasted work.

However, I don't _think_ it's a significant DDoS vector, because it looks like the call to `get_or_create_user` only happens right at the end of the OIDC backend's [authenticate method](https://github.com/mozilla/mozilla-django-oidc/blob/71e4af8283a10aa51234de705d34cd298e927f97/mozilla_django_oidc/auth.py#L297C29-L297C47), and to reach there, my reading of the code says that the user has to be already proven as having a legitimate @mozilla.com identity. If this holds true - and please do review the auth backend linked above) the risk of someone spamming the login flow is restricted to Mozilla staff (or someone who has siezed such an account), which I think is pretty low.

Also, we have rate-limiting in the MVP backlog, which will further mitigate this.

Thoughts (and sense-checking) very welcome here, please 🙇

Reply:
I agree with your reading that it attempts to authenticate and verify credentials before attempting to create a user. 👍

Reply:
Thanks for the exploration @robhudson. I had to bootstrap a non-SSO user while infra was set up and it was actually easy enough to make me prefer the longhand method, after all, rather than add the security risk (and env var is all is takes to slide in a new admin) and wasted DB hits. So I think i'll add a commit to pull it all out before i merge, after all